shared/src/main/scala/cache.scala

```diff
@@ -16,55 +16,32 @@
 
 package fetch
 
+import cats.effect._
+
 /**
  * A `Cache` trait so the users of the library can provide their own cache.
  */
 trait DataSourceCache {
-  def update[A](k: DataSourceIdentity, v: A): DataSourceCache
-
-  def get[A](k: DataSourceIdentity): Option[A]
-
-  def getWithDS[I, A](ds: DataSource[I, A]): I => Option[A] = id => get(ds.identity(id))
-
-  def cacheResults[I, A](results: Map[I, A], ds: DataSource[I, A]): DataSourceCache = {
-    results.foldLeft(this)({
-      case (acc, (i, a)) => acc.update(ds.identity(i), a)
-    })
-  }
-
-  def contains(k: DataSourceIdentity): Boolean = get(k).isDefined
+  def lookup[I, A](i: I, ds: DataSource[I, A]): IO[Option[A]]
+  def insert[I, A](i: I, ds: DataSource[I, A], v: A): IO[DataSourceCache]
 }
 
 /**
  * A cache that stores its elements in memory.
  */
-case class InMemoryCache(state: Map[DataSourceIdentity, Any]) extends DataSourceCache {
-  override def get[A](k: DataSourceIdentity): Option[A] =
-    state.get(k).asInstanceOf[Option[A]]
+case class InMemoryCache(state: Map[(String, Any), Any]) extends DataSourceCache {
```

Comment:
`(String, Any)` represents the DataSource name and the "key", right?

Can we define both with a [value classes](https://docs.scala-lang.org/overviews/core/value-classes.html)?
```scala
final class DataSourceName(val name: String) extends AnyVal
final class DataSourceKey(val key: Any) extends AnyVal
```